presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1326,11 +1327,34 @@ protected Scope visitJoin(Join node, Optional<Scope> scope)
             Scope left = process(node.getLeft(), scope);
             Scope right = process(node.getRight(), isLateralRelation(node.getRight()) ? Optional.of(left) : scope);
 
-            if (isLateralRelation(node.getRight()) && (node.getType().equals(RIGHT) || node.getType().equals(FULL))) {
-                Stream<Expression> leftScopeReferences = getReferencesToScope(node.getRight(), analysis, left);
-                leftScopeReferences.findFirst().ifPresent(reference -> {
-                    throw semanticException(INVALID_COLUMN_REFERENCE, reference, "LATERAL reference not allowed in %s JOIN", node.getType().name());
-                });
+            if (isLateralRelation(node.getRight())) {
+                if (node.getType().equals(RIGHT) || node.getType().equals(FULL)) {
+                    Stream<Expression> leftScopeReferences = getReferencesToScope(node.getRight(), analysis, left);
+                    leftScopeReferences.findFirst().ifPresent(reference -> {
+                        throw semanticException(INVALID_COLUMN_REFERENCE, reference, "LATERAL reference not allowed in %s JOIN", node.getType().name());
+                    });
+                }
+                if (isUnnestRelation(node.getRight())) {
+                    if (criteria != null) {
```

Comment:
This null check is unnecessary. If “criteria” is null, “criteria instanceof JoinOn” will be false

Reply:
Join involving Unnest is supported both without criteria (cross join) and with criteria ON TRUE. If I remove the null check, cross join will no longer be accepted.